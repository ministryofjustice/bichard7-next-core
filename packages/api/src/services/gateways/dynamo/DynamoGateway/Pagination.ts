import type DynamoDB from "aws-sdk/clients/dynamodb"

export default interface Pagination {
  lastItemKey?: DynamoDB.DocumentClient.Key
  limit?: number
}
