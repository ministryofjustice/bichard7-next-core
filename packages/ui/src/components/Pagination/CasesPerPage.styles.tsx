import { SelectInput } from "@govuk-react/select"
import styled from "styled-components"

const CasesPerPagePicker = styled(SelectInput)`
  width: auto;
`

export { CasesPerPagePicker }
