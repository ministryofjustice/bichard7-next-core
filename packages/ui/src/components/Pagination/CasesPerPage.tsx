import { useRouter } from "next/router"
import { CasesPerPagePicker } from "./CasesPerPage.styles"

interface Props {
  options: number[]
  selected: number
  pageNum: number
  casesPerPage: number
}

const CasesPerPage: React.FC<Props> = ({ options, selected, pageNum, casesPerPage }: Props) => {
  const router = useRouter()

  return (
    <div className="moj-pagination__results">
      {"View "}
      <CasesPerPagePicker
        onChange={(event) => {
          const newCasesPerPage = event.target.value

          // Ensure that the first case on the page remains after changing number of cases per page
          const firstCaseIndex = (pageNum - 1) * casesPerPage + 1
          const newPageNum = Math.floor(firstCaseIndex / parseInt(newCasesPerPage, 10)) + 1

          router.push({ query: { ...router.query, maxPageItems: newCasesPerPage, page: newPageNum } })
        }}
        value={selected}
        className={`cases-per-page cases-per-page-picker`}
        aria-label="Cases per page"
      >
        {options.map((option) => (
          <option value={option} key={option}>
            {option}
          </option>
        ))}
      </CasesPerPagePicker>
      {" cases per page"}
    </div>
  )
}

export default CasesPerPage
