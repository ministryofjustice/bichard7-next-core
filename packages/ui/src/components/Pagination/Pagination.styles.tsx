import styled from "styled-components"

const PaginationBar = styled.div`
  display: inline-flex;
  flex-wrap: nowrap;
  justify-content: space-between;
  align-items: center;
  width: 99%;
`

export { PaginationBar }
