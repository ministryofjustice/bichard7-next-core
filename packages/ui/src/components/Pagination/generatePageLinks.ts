export type PageLabel = "Previous" | number | "Ellipsis" | "Next"
export type PageLink = { label: PageLabel; destinationPage?: number }

// Returns an list of elements to display in the pagination widget
// Page numbers are 1-indexed
// totalPages should be >= 0 and currentPage should be >=0 and <= totalPages
export const generatePageLinks = (currentPage: number, totalPages: number): PageLink[] => {
  const labels: PageLink[] = []

  // Don't show any pagination controls if there's only one page
  if (totalPages < 2) {
    return []
  }

  // Previous page arrow
  if (currentPage > 1) {
    labels.push({
      label: "Previous",
      destinationPage: currentPage - 1
    })
  }

  // First page
  if (currentPage > 2) {
    labels.push({ label: 1, destinationPage: 1 })
  }

  // Abridged gap between first page and previous page
  if (currentPage > 3) {
    labels.push({ label: "Ellipsis" })
  }

  // Previous page
  if (currentPage > 1) {
    labels.push({ label: currentPage - 1, destinationPage: currentPage - 1 })
  }

  // Current page
  if (currentPage > 0) {
    labels.push({ label: currentPage })
  }

  // Next page
  if (currentPage < totalPages) {
    labels.push({ label: currentPage + 1, destinationPage: currentPage + 1 })
  }

  // Abridged gap between next and last page
  if (currentPage < totalPages - 2) {
    labels.push({ label: "Ellipsis" })
  }

  // Last page
  if (currentPage < totalPages - 1) {
    labels.push({ label: totalPages, destinationPage: totalPages })
  }

  // Next page arrow
  if (currentPage < totalPages) {
    labels.push({ label: "Next", destinationPage: currentPage + 1 })
  }
  return labels
}
