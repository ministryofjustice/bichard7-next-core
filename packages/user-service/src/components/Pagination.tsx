import addQueryParams from "utils/addQueryParams"
import Link from "./Link"

interface Props {
  pageNumber: number
  totalItems: number
  maxItemsPerPage: number
  href: string
  className?: string
}

const Pagination = ({ pageNumber, totalItems, maxItemsPerPage, href, className }: Props) => {
  if (totalItems === 0) {
    return <></>
  }

  const nextPage = addQueryParams(href, {
    page: pageNumber + 1
  })

  const prevPage = addQueryParams(href, {
    page: pageNumber - 1
  })

  const pageNumberStyle = {
    style: {
      padding: "10px"
    },
    spacing: {
      paddingLeft: "0px"
    }
  }

  let styles = { ...pageNumberStyle.style }

  if (pageNumber === 0) {
    styles = { ...styles, ...pageNumberStyle.spacing }
  }

  const pageString = `Page ${pageNumber + 1} of ${Math.ceil(totalItems / maxItemsPerPage)}`

  return (
    <div className={`govuk-hint${className ? ` ${className}` : ""}`}>
      <Link href={prevPage} data-test="Prev">
        {pageNumber > 0 && "< Prev"}
      </Link>
      <span style={styles}>{pageString}</span>
      <Link href={nextPage} data-test="Next">
        {pageNumber + 1 < totalItems / maxItemsPerPage && "Next >"}
      </Link>
    </div>
  )
}

export default Pagination
