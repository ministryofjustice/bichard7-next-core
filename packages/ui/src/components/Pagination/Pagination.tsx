import defaults from "defaults"
import ConditionalRender from "../ConditionalRender"
import CasesPerPage from "./CasesPerPage"
import { PaginationBar } from "./Pagination.styles"
import PaginationNavigation from "./PaginationNavigation"
import PaginationResults from "./PaginationResults"

interface Props {
  pageNum?: number
  casesPerPage?: number
  totalCases: number
  name?: string
}

const Pagination: React.FC<Props> = ({
  pageNum = 1,
  casesPerPage = defaults.maxPageItems,
  totalCases,
  name
}: Props) => {
  return (
    <ConditionalRender isRendered={totalCases > 0}>
      <PaginationBar id={`${name}-pagination-bar`} className={"pagination-bar"}>
        <PaginationResults pageNum={pageNum} casesPerPage={casesPerPage} totalCases={totalCases} />
        <CasesPerPage
          pageNum={pageNum}
          casesPerPage={casesPerPage}
          options={[25, 50, 100, 200]}
          selected={casesPerPage}
        />
        <PaginationNavigation pageNum={pageNum} totalPages={Math.ceil(totalCases / casesPerPage)} name={name} />
      </PaginationBar>
    </ConditionalRender>
  )
}

export default Pagination
