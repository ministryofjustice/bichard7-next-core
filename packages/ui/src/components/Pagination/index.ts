import Pagination from "./Pagination"
import PaginationNavigation from "./PaginationNavigation"
import PaginationResults from "./PaginationResults"
import CasesPerPage from "./CasesPerPage"

export default Pagination

export { PaginationResults, CasesPerPage, PaginationNavigation }
