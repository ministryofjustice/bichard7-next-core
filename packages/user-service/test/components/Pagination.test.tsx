import { render } from "@testing-library/react"
import Pagination from "components/Pagination"

it("should render the component and match the snapshot when total items is more than zero", () => {
  const { container } = render(<Pagination href="/dummy-url" maxItemsPerPage={10} pageNumber={1} totalItems={100} />)

  expect(container).toMatchSnapshot()
})

it("should render the component and match the snapshot when total items is zero", () => {
  const { container } = render(<Pagination href="/dummy-url" maxItemsPerPage={10} pageNumber={1} totalItems={0} />)

  expect(container).toMatchSnapshot()
})
