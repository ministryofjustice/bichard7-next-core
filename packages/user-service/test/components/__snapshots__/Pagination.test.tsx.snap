// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`should render the component and match the snapshot when total items is more than zero 1`] = `
<div>
  <div
    class="govuk-hint"
  >
    <a
      class="govuk-link"
      data-test="Prev"
      href="/dummy-url?page=0"
    >
      &lt; Prev
    </a>
    <span
      style="padding: 10px;"
    >
      Page 2 of 10
    </span>
    <a
      class="govuk-link"
      data-test="Next"
      href="/dummy-url?page=2"
    >
      Next &gt;
    </a>
  </div>
</div>
`;

exports[`should render the component and match the snapshot when total items is zero 1`] = `<div />`;
